### PR TITLE
add more flexibility for shutdown time

### PR DESCRIPTION
**Pull request by mwhudson**
This branch adds two parameters to the shutdown (renamed from reboot) endpoint:

 * 'immediate' for rebooting right now, not after the install / postinstall steps are complete
 * a mode to choose between rebooting and powering off

cc @jpnurmi

### CODE REVIEW

**Comment by mwhudson** [comment]
For https://bugs.launchpad.net/subiquity/+bug/1940263 and https://bugs.launchpad.net/subiquity/+bug/1940441

**Comment by jpnurmi** [comment]
Thanks. Works for me

File: subiquity/server/controllers/shutdown.py [hunk 1]
```diff
@@ -23,42 +23,65 @@
...
     def start(self):
+        self.app.aio_loop.create_task(self._wait_install())
         self.app.aio_loop.create_task(self._run())
 
-    async def _run(self):
-        Install = self.app.controllers.Install
-        await Install.install_task
+    async def _wait_install(self):
+        await self.app.controllers.Install.install_task
+        self.install_complete = True
```
Comment (dbungert) [inline subiquity/server/controllers/shutdown.py:73]: To my eyes, install_complete is redundant with `self.app.state == ApplicationState.DONE` - am I missing something?

File: subiquity/server/controllers/shutdown.py [hunk 2]
```diff
@@ -23,42 +23,65 @@
...
+    autoinstall_schema = {
+        'type': 'string',
+        'enum': ['reboot', 'poweroff']
+    }
 
     def __init__(self, app):
         super().__init__(app)
-        self.user_reboot_event = asyncio.Event()
-        self.rebooting_event = asyncio.Event()
+        self.user_shutdown_event = asyncio.Event()
```
Comment (dbungert) [inline subiquity/server/controllers/shutdown.py:44]: I think the event names are similar enough, and nuanced enough in their concept, to be worthy of comments explaining their intent.  Let me have a try
user_shutdown_event # the user has requested the shutdown
server_reboot_event # the server is ready for shutdown
shuttingdown_event # the actual shutdown is about to start

**Review by dbungert — CHANGES_REQUESTED** [review]

File: subiquity/common/types.py [hunk 3]
```diff
@@ -330,3 +330,8 @@ class SnapListResponse:
 class TimeZoneInfo:
     timezone: str
     from_geoip: bool
+
+
+class ShutdownMode(enum.Enum):
+    REBOOT = enum.auto()
+    POWER_OFF = enum.auto()
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/common/types.py:337]: I mean, sure :)

File: subiquity/server/controllers/shutdown.py (hunk 1 above)
Comment (mwhudson) [reply subiquity/server/controllers/shutdown.py:73]: Probably not, no.

**Comment by mwhudson** [comment]
Thanks for the review, I've pushed the fixes as !fixup commits which I'll rebase before merge.

**Review by dbungert — APPROVED** [review]
I really like the fixup commits.  They make it easy to see what has changed, while still making it obvious what to do later with `rebase -i`

**Pull request merged**